account-keys/slip10/src/lib.rs

```diff
@@ -0,0 +1,434 @@
+use mc_crypto_keys::RistrettoPrivate;
+use sha2::Sha512;
+use zeroize::Zeroize;
+
+/// An enumeration of errors which can occur while working with SLIP-0010 key
+/// derivation
+#[derive(Debug, Display, Eq, PartialEq)]
+pub enum Error {
+    /// The path provided contained a member that was not "hardened".
+    // FIXME: this is currently unused, and the slip10_ed25519 crate clamps to the appropriate
+    //        range.
+    UnhardenedPath,
+    /// There was an error creating the account key: {0}
+    AccountKey(AccountKeyError),
+}
+
+/// The result type
+pub type Result<T> = CoreResult<T, Error>;
+
+/// A key derived using SLIP-0010 key derivation
+#[derive(Zeroize)]
+#[zeroize(drop)]
+pub struct Slip10Key([u8; 32]);
+
+impl AsRef<[u8]> for Slip10Key {
+    fn as_ref(&self) -> &[u8] {
+        &self.0[..]
+    }
+}
+
+/// Create the view and spend private keys, and return them in reverse order,
+/// e.g. `(spend, view)`, to match `AccountKey::new()`
+impl Into<(RistrettoPrivate, RistrettoPrivate)> for Slip10Key {
+    fn into(self) -> (RistrettoPrivate, RistrettoPrivate) {
+        let mut okm = [0u8; 64];
+
+        let view_kdf = Hkdf::<Sha512>::new(Some(b"mobilecoin-ristretto255-view"), self.as_ref());
+        view_kdf
+            .expand(b"", &mut okm)
+            .expect("Invalid okm length when creating private view key");
+        let view_scalar = Scalar::from_bytes_mod_order_wide(&okm);
+        let view_private_key = RistrettoPrivate::from(view_scalar);
+
+        let spend_kdf = Hkdf::<Sha512>::new(Some(b"mobilecoin-ristretto255-spend"), self.as_ref());
+        spend_kdf
+            .expand(b"", &mut okm)
+            .expect("Invalid okm length when creating private spend key");
+        let spend_scalar = Scalar::from_bytes_mod_order_wide(&okm);
+        let spend_private_key = RistrettoPrivate::from(spend_scalar);
+
+        (spend_private_key, view_private_key)
+    }
+}
+
+impl From<[u8; 32]> for Slip10Key {
+    fn from(src: [u8; 32]) -> Self {
+        Self(src)
+    }
+}
+
+/// A common interface for constructing a Slip10Key at a particular path from
+/// existing entropy
+pub trait Slip10KeyGenerator {
+    /// The type of error, if any, to be returned if it occurs
+    type Error: Display;
+
+    /// Derive a slip10 key for the given path from the current object
+    fn derive_slip10_key(self, path: &[u32]) -> CoreResult<Slip10Key, Self::Error>;
+}
+
+// This lets us get to
+// Mnemonic::from_phrases().derive_slip10_key(path).try_into_account_key(...)
+impl Slip10KeyGenerator for Mnemonic {
+    type Error = Error;
+
+    fn derive_slip10_key(self, path: &[u32]) -> Result<Slip10Key> {
+        // We explicitly do not support passphrases for BIP-39 mnemonics, please
+        // see the Mobilecoin Key Derivation design specification, v1.0.0, for
+        // design rationale.
+        let seed = Seed::new(&self, "");
+        let key = slip10_ed25519::derive_ed25519_private_key(seed.as_bytes(), path);
```

Comment:
Is this name `derive_ed25519_private_key()` misleading or am I crazy? AFAICT it just produces a pseudorandom 256 bit number, calling it an 'ed25519 private key' is generous (it made me think it was reduced mod the group order - which is important in this context where the slip10 key will be hashed).

If I am not crazy it would be nice to add a comment explaining what `derive_ed25519_private_key()` actually does.

Reply:
The [SLIP-0010 specification itself](https://github.com/satoshilabs/slips/blob/master/slip-0010.md#master-key-generation) labels this activity as "generating a private key," because "every 256-bit number (even 0) is a valid private key" which is (I think) why Alan named that function as he did. I'm fine to comment that this isn't taking the entropy across the finish line, tho ;-)

Reply:
Yeah the SLIP-0010 specification is technically wrong when it says that any 256-bit number is a valid private key.